### PR TITLE
SC: ExtraData is added into request/handle method.

### PR DESCRIPTION
**Pull request by KimKyungup**
## Proposed changes
This PR added extraData argument into request/handle method.
This change is needed for SP to develop their own extended Bridge contract based on the Bridge contract. So extraData is added like general form.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues
More unit test will be added in next PR.
Also, klaytn-node-test will be modified to support this.

## Related comment
Dear reviewers. :)
This PR is following PR of #115 .
So Please review the below commit before #115 is merged.
5cfeeab

I will rebase this later.

### CODE REVIEW

**Comment by KimKyungup** [comment]
Dear reviewers. :)
This PR is following PR of #115 .
So Please review the below commit before #115 is merged.
https://github.com/klaytn/klaytn/pull/117/commits/5cfeeabb13a9f59c4068f7ea45080a0174255190

I will rebase this later.

**Comment by KimKyungup** [comment]
After previous PR is merged, I will change target branch to master. :) PTAL.

File: contracts/bridge/Bridge.sol [hunk 1]
```diff
@@ -63,7 +63,8 @@ contract Bridge is IERC20BridgeReceiver, IERC721BridgeReceiver, BridgeFee, Bridg
         uint256 valueOrTokenId,
         uint64 requestNonce,
         string uri,
-        uint256 fee
+        uint256 fee,
+        uint256 []extraData
```
Comment (aidan-kwon) [inline contracts/bridge/Bridge.sol]:
```suggestion
        uint256[] extraData
```
https://solidity.readthedocs.io/en/v0.5.10/style-guide.html#variable-declarations

File: contracts/bridge/Bridge.sol [hunk 2]
```diff
@@ -83,7 +84,9 @@ contract Bridge is IERC20BridgeReceiver, IERC721BridgeReceiver, BridgeFee, Bridg
         address to,
         address tokenAddress,
         uint256 valueOrTokenId,
-        uint64 handleNonce
+        uint64 handleNonce,
+        uint256 []extraData
```
Comment (aidan-kwon) [inline contracts/bridge/Bridge.sol]: https://solidity.readthedocs.io/en/v0.5.10/style-guide.html#variable-declarations

File: contracts/bridge/Bridge.sol [hunk 3]
```diff
@@ -229,24 +262,33 @@ contract Bridge is IERC20BridgeReceiver, IERC721BridgeReceiver, BridgeFee, Bridg
...
-    function requestKLAYTransfer(address _to, uint256 _value) external payable {
+    function requestKLAYTransfer(address _to, uint256 _value, uint256 []_extraData) external payable {
         uint256 feeLimit = msg.value.sub(_value);
-        _requestKLAYTransfer(_to, feeLimit);
+        _requestKLAYTransfer(_to, feeLimit, _extraData);
     }
 
     // _requestERC20Transfer requests transfer ERC20 to _to on relative chain.
-    function _requestERC20Transfer(address _tokenAddress, address _from, address _to, uint256 _value, uint256 _feeLimit) internal {
+    function _requestERC20Transfer(address _tokenAddress,
```
Comment (aidan-kwon) [inline contracts/bridge/Bridge.sol]:
```suggestion
    function _requestERC20Transfer(
        address _tokenAddress,
```

File: contracts/bridge_test.go [hunk 4]
```diff
@@ -437,23 +450,59 @@ func TestExtendedBridgeAndCallback(t *testing.T) {
...
+		assert.Equal(t, erc20Addr, ev.TokenAddress)
+		assert.Equal(t, sc.ERC20, ev.TokenType)
+		assert.Equal(t, bobAcc.From, ev.To)
+		assert.Equal(t, len(extraData), len(ev.ExtraData))
+		assert.Equal(t, extraData[0].String(), ev.ExtraData[0].String())
+		assert.Equal(t, extraData[1].String(), ev.ExtraData[1].String())
+
+		// HandleERC20Transfer
+		tx, err = b.HandleERC20Transfer(bridgeAccount, ev.Raw.TxHash, ev.From, ev.To, ev.TokenAddress, ev.ValueOrTokenId, ev.RequestNonce, ev.Raw.BlockNumber, ev.ExtraData)
+		assert.Equal(t, nil, err)
```
Comment (jeonghonah) [inline contracts/bridge_test.go]: assert.NoError(t, err)
```suggestion
		assert.NoError(t, err)
```

File: contracts/bridge_test.go [hunk 5]
```diff
@@ -437,23 +450,59 @@ func TestExtendedBridgeAndCallback(t *testing.T) {
...
-	assert.Equal(t, nil, err)
+	amount := big.NewInt(1000)
+	offerPrice := big.NewInt(10)
+	dummyBigInt := big.NewInt(20)
+	extraData := []*big.Int{offerPrice, dummyBigInt}
+
+	tx, err = erc20.Approve(aliceAcc, bridgeAddr, amount)
+	assert.NoError(t, err)
+	backend.Commit()
+	assert.Nil(t, WaitMined(tx, backend, t))
```
Comment (jeonghonah) [inline contracts/bridge_test.go]: How about using `CheckWaitMined()`?

File: contracts/extbridge/ext_bridge.sol [hunk 6]
```diff
@@ -21,19 +21,17 @@ contract ExtBridge is Bridge {
...
         uint64 _requestNonce,
-        uint64 _requestBlockNumber
-        //uint256 [] _extraData     // TODO-Klaytn-Servicechain : This will be applied in next PR
+        uint64 _requestBlockNumber,
+        uint256 [] _extraData
     )
         public
     {
-        uint256 offerPrice = 1; //_extraData[0];
+        uint256 offerPrice = _extraData[0];
```
Comment (jeonghonah) [inline contracts/extbridge/ext_bridge.sol]: What happen if _extraData is nil?

Comment (KimKyungup) [reply contracts/extbridge/ext_bridge.sol]: It makes revert. I think adding length check code is needed.

**Pull request merged**